frontend/src/metabase/visualizations/lib/apply_tooltips.js

```diff
@@ -6,160 +6,207 @@ import d3 from "d3";
+function setHoverFunctions(chart, series, isStacked, isNormalized, onHoverChange) {
+    let [{ data: { cols } }] = series;
+    chart.selectAll(".bar, .dot, .area, .line, .bubble")
+         .on("mousemove", function(d, i) {
+             const seriesIndex       = determineSeriesIndexFromElement(this, isStacked);
+             const thisSeries        = series[seriesIndex];
+             const card              = thisSeries.card;
+             const isSingleSeriesBar = this.classList.contains("bar") && series.length === 1;
+             const isArea            = this.classList.contains("area");
+             const isScatter         = Array.isArray(d.key);
+
+             let data = [];
+             if (isScatter) {
+                 data = getHoverDataForScatterChart(d, cols);
+             }
+             // line, area, bar
+             else if (d.data) {
+                 if (!isSingleSeriesBar) {
+                     cols = thisSeries.data.cols;
+                 }
+                 // !!!! for some reason i starts at 1 instead of 0 !!!!
+                 data = getHoverDataForLineAreaBarChart(d, i - 1, thisSeries, cols, isNormalized);
+             }
+
+             // not sure what we're doing here or why
+             if (data && series.length > 1) {
+                 if (card._breakoutColumn) {
+                     data.unshift({
+                         key: getFriendlyName(card._breakoutColumn),
+                         value: card._breakoutValue,
+                         col: card._breakoutColumn
                      });
-                 })
-                 .on("mouseleave", function() {
-                     if (!onHoverChange) {
-                         return;
-                     }
-                     onHoverChange(null);
-                 })
+                 }
+             }
+
+             // remove any duplicate columns
+             data = _.uniq(data, (d) => d.col);
+
+             onHoverChange({
+                 // for single series bar charts, fade the series and highlght the hovered element with CSS
+                 index: isSingleSeriesBar ? -1 : seriesIndex,
+                 // for area charts, use the mouse location rather than the DOM element
+                 element: isArea ? null : this,
+                 event: isArea ? d3.event : null,
+                 data: data.length > 0 ? data : null,
+             });
+         })
+         .on("mouseleave", function() {
+             onHoverChange(null);
+         });
+}
+
+function setClickFunctions(chart, series, isStacked, isNormalized, isScalarSeries, onVisualizationClick) {
+    let [{ data: { cols } }] = series;
+    const onClick = function(d) {
+        const seriesIndex       = determineSeriesIndexFromElement(this, isStacked);
+        const thisSeries        = series[seriesIndex];
+        const card              = thisSeries.card;
+        const isSingleSeriesBar = this.classList.contains("bar") && series.length === 1;
```

Comment:
It's a bit of a bummer we have to determine these properties again here just like we have to in the `setHoverFunctions` fn.

Reply:
My thinking here was I was trying to minimize the number of params that get passed to the newly-broken-out functions. Things like `isSingleSeriesBar` to me is an implict property of the chart/series so I didn't see the need to make it a separate param. But it would probably be a good idea to move the logic for determining that into a separate function so it doesn't have to be repeated. Because twice as much logic = twice as many places bugs can happen